### PR TITLE
add a lock to make LOG_HEAD thread-safe #{317}

### LINKED ISSUE #317: [BUG] localtime is not thread-safe
**Describe the bug**
`localtime` is used in LOG_HEAD but it is not a thread-safe function.

### PR DESCRIPTION
**Pull request by Wenbin1002**
### What problem were solved in this pull request?

Issue Number: close #317 

Problem:

### CODE REVIEW

File: deps/common/log/log.cpp [hunk 1]
```diff
@@ -312,6 +312,18 @@ void Log::set_context_getter(std::function<intptr_t()> context_getter)
 
 intptr_t Log::context_id() { return context_getter_(); }
 
+void Log::get_time_lock()
+{
+  pthread_mutex_init(&get_time_lock_, NULL);
```
Comment (hnwyllmm) [inline deps/common/log/log.cpp:317]: 1. if you init a lock every time, then the lock cannot protect the critical section;
2. get_time_lock maybe called in parallel, and `pthread_mutex_init` is not thread-safe;
3. std::mutex is easier.

File: deps/common/log/log.h [hunk 2]
```diff
@@ -185,7 +189,9 @@ extern Log *g_log;
   if (common::g_log) {                                                     \
     struct timeval tv;                                                     \
     gettimeofday(&tv, NULL);                                               \
+    common::g_log->get_time_lock();                                        \
     struct tm *p                      = localtime(&tv.tv_sec);             \
```
Comment (hnwyllmm) [inline deps/common/log/log.h:193]: localtime is not thread-safe but there is a thread-safe interface called localtime_r. You can refer to manual book to check it.

**Review by hnwyllmm — CHANGES_REQUESTED** [review]
Thanks for your contribution. But there are some problems should be resolved.

File: deps/common/log/log.h (hunk 2 above)
Comment (Wenbin1002) [reply deps/common/log/log.h:193]: So just change localtime to localtime_r, no need for additional lock?

Comment (hnwyllmm) [reply deps/common/log/log.h:193]: That's the easiest method.
But localtime_r has bad performance, so it's welcomed if you have better solution.

**Review by hnwyllmm — APPROVED** [review]

**Comment by hnwyllmm** [comment]
I edit your file directly by clang-format.
When I tried to create a pull request, an error occurs so I give up.

**Comment by Wenbin1002** [comment]
> I edit your file directly by clang-format. When I tried to create a pull request, an error occurs so I give up.

sry, my first time to contribute, I'll check next time.

**Comment by hnwyllmm** [comment]
That's fine. Thanks for your contribution.

**Pull request merged**